gobblin-utility/src/main/java/gobblin/util/ExecutorsUtils.java

```diff
@@ -146,4 +154,68 @@ public static void shutdownExecutorService(ExecutorService executorService, Opti
     shutdownExecutorService(executorService, logger, EXECUTOR_SERVICE_SHUTDOWN_TIMEOUT,
         EXECUTOR_SERVICE_SHUTDOWN_TIMEOUT_TIMEUNIT);
   }
+
+  /**
+   * A utility method to parallelize loops. Applies the {@link Function} to every element in the {@link List} in
+   * parallel by spawning threads. A list containing the result obtained by applying the function is returned. The
+   * method is a blocking call and will wait for all the elements in the list to be processed or timeoutInSecs which
+   * ever is earlier.
+   *
+   * <ul>
+   * <li>Uses a Fixed thread pool of size threadCount.
+   * <li>Uses {@link #shutdownExecutorService(ExecutorService, Optional, long, TimeUnit)} to shutdown the executor
+   * service
+   * <li>All threads are daemon threads
+   * </ul>
+   *
+   * @param list input list on which the function is applied in parallel
+   * @param function to be applied on every element of the list
+   * @param threadCount to be used to process the list
+   * @param timeoutInSecs to wait for all the threads to complete
+   * @param logger an {@link Optional} wrapping a {@link Logger} to be used during shutdown
+   *
+   * @return a list containing the result obtained by applying the function on each element of the input list in the
+   *         same order
+   *
+   * @throws IllegalArgumentException if input list or function is null
+   */
+  public static <T, V> List<T> parallelize(List<V> list, final Function<V, T> function, int threadCount,
```

Comment:
1) I think inputOutput is unnecessary. Your task logic can be along the lines of result.set(i, function(list.get(i)) if result is pre-allocated.

2) You need to handle errors in function() or timeouts. You need to use submit() instead of execute() and store all Futures. Once all tasks are done, you need to check if all tasks completed successfully. If any of them failed or timed out, parallelize() needs to through an exception.